### ISSUE TITLE
Add support for ports other than 80/443

### ISSUE THREAD

**Issue by soggiest** [heptio/contour#661]
I would like to see contour support ports outside of 80 and 443 for Virtualservices and Routes. Allowing virtualservices to only be created for two ports is highly restrictive. Looking at the source code it looks like ports 80 and 443 are hard coded, variables could easily be used for ports and allow for greater flexibility.

Use case:
I have a bare-metal K8s environment sitting behind a firewall. Firewall rules/portforwarding for ports 80 and 443 are restricted as many people are using the network behind the firewall. I've enabled port forwarding from an unused port ("30000" in this instance) to the Contour loadbalancer. When I try to route to my server from outside the firewall I have to include port 30000 to get the routing to work properly, but this will set my `Host` header to be `FQDN:30000`. Because of the way the `Host` header is set I cannot route to my service via contour.

**Comment by davecheney** [comment]
Hi there. Thanks for logging this issue. There is some limited flexbility for controlling the listening ports in contour at the moment (see contour bootstrap help), but you are correct that the insecure port is pinned to 80 and the secure port is pinned to 443. 

I also want this because I would like to move some of the configuration out of the hard coded bootstrap config into configuration that is automatically inserted into the LDS/RDS configuration sent to the running envoy. I think I know how to do this at the level of `internal/contour` creating its own entries in the DAG, but I'm not sure how to expose that to our ingress and ingressroute documents. The obvious answer is yet more annotations on the ingress document (which you can probably infer from my tone, is not something I really want to do). There is perhaps some more scope for adding this in our ingressroute document; we could add a port field to the `spec.virutalhost` and `spec.virtualhost.tls` stanzas.

I'm not going to schedule this issue to a release because I don't have a clear vision for how it would be implemented.

**Comment by stevesloka** [comment]
I'm thinking we can add some args to Contour to tell envoy which ports VirtualHosts are configured for. Doing this we would be able to have Contour listen on the same port that Envoy is configured to listen on. 

```
--envoy-http-virtualhost-port
--envoy-https-virtualhost-port
```

// cc @joshrosso

**Comment by moderation** [comment]
@stevesloka and @joshrosso are aware of our use case which is allowing us to run Contour in parallel to another non-Contour ingress controller for testing purposes (primarily TCP ingress). However the ability to run multiple Ingress controllers might be a production requirement until (or if) Contour can expose Envoy functionality like sticky sessions for cookies and external authorization.

**Comment by davecheney** [comment]
After thinking about this problem for a while I'm not sure adding flags to change this is the right solution. The underlying issue is Envoy considers these two headers `Host: example.com` and `Host: example.com:80` to be _different_ virtual hosts. This is why requests to http://example.com:5000, in the example above, fail. 

The current workaround, untested, is to tell Contour to configure Envoy to listen on the same port that is TCP forwarded via `--envoy-http{,s}--port`. The longer term fix is envoyproxy/envoy#1269
